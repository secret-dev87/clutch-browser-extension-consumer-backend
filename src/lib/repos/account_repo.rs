use sea_orm::Set;
use sea_orm::{entity::prelude::*, sea_query::Expr};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Serialize, Deserialize)]
#[sea_orm(table_name = "accounts")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: String,
    pub email: String,
    pub wallet_address: String,
    pub eoa_address: String,
    pub eoa_private_address: String,
    pub updated_at: i64,
}

#[derive(Copy, Clone, Debug, EnumIter)]
pub enum Relation {}

impl RelationTrait for Relation {
    fn def(&self) -> RelationDef {
        panic!("No RelationDef")
    }
}

impl ActiveModelBehavior for ActiveModel {}

pub async fn create(
    db: &DatabaseConnection,
    id: Uuid,
    email: String,
    wallet_address: String,
    eoa_address: String,
    eoa_private: String,
    updated_at: i64,
) -> anyhow::Result<()> {
    let model = ActiveModel {
        id: Set(id.to_string()),
        email: Set(email.to_owned()),
        wallet_address: Set(wallet_address.to_owned()),
        eoa_address: Set(eoa_address.to_owned()),
        eoa_private_address: Set(eoa_private.to_owned()),
        updated_at: Set(updated_at.to_owned()),
    };

    Entity::insert(model)
        .exec_without_returning(db)
        .await
        .map(|_| ())
        .map_err(|e| anyhow::anyhow!(e))
}

pub async fn find_by_id(
    db: &DatabaseConnection,
    account_id: String,
) -> anyhow::Result<Option<Model>> {
    Entity::find()
        .filter(Column::Id.eq(account_id))
        .one(db)
        .await
        .map_err(|e| anyhow::anyhow!(e))
}

pub async fn find_by_email(db: &DatabaseConnection, email: &str) -> anyhow::Result<Option<Model>> {
    Entity::find()
        .filter(Column::Email.eq(email))
        .one(db)
        .await
        .map_err(|e| anyhow::anyhow!(e))
}

pub async fn find_by_wallet_address(
    db: &DatabaseConnection,
    wallet_address: String,
) -> anyhow::Result<Option<Model>> {
    Entity::find()
        .filter(Column::WalletAddress.eq(wallet_address))
        .one(db)
        .await
        .map_err(|e| anyhow::anyhow!(e))
}

pub async fn find_all_by_wallet_address(
    db: &DatabaseConnection,
    wallet_address: String,
) -> anyhow::Result<Vec<Model>> {
    Entity::find()
        .filter(Column::WalletAddress.eq(wallet_address))
        .all(db)
        .await
        .map_err(|e| anyhow::anyhow!(e))
}

pub async fn find_all_by_eoa_address(
    db: &DatabaseConnection,
    eao_address: String,
) -> anyhow::Result<Vec<Model>> {
    Entity::find()
        .filter(Column::EoaAddress.eq(eao_address))
        .all(db)
        .await
        .map_err(|e| anyhow::anyhow!(e))
}

pub async fn find_all_by_email_address(
    db: &DatabaseConnection,
    email: String,
) -> anyhow::Result<Vec<Model>> {
    Entity::find()
        .filter(Column::Email.eq(email))
        .all(db)
        .await
        .map_err(|e| anyhow::anyhow!(e))
}

pub async fn find_all_by_account_ids(
    db: &DatabaseConnection,
    account_ids: Vec<String>,
) -> anyhow::Result<Vec<Model>> {
    Entity::find()
        .filter(Column::Id.is_in(account_ids))
        .all(db)
        .await
        .map_err(|e| anyhow::anyhow!(e))
}

pub async fn find_all(db: &DatabaseConnection) -> anyhow::Result<Vec<Model>> {
    Entity::find().all(db).await.map_err(|e| anyhow::anyhow!(e))
}

pub async fn update(
    db: &DatabaseConnection,
    id: String,
    wallet_address: Option<String>,
    eoa_address: Option<String>,
) -> anyhow::Result<()> {
    Entity::update_many()
        .col_expr(
            Column::WalletAddress,
            Expr::value(wallet_address.unwrap_or_default()),
        )
        .col_expr(
            Column::EoaAddress,
            Expr::value(eoa_address.unwrap_or_default()),
        )
        .filter(Column::Id.eq(id))
        .exec(db)
        .await
        .map_err(|e| anyhow::anyhow!(e))?;
    Ok(())
}
