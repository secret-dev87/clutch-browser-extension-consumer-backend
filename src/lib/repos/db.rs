use sea_orm::entity::prelude::*;
use sea_orm::Database;

use crate::config::settings::Settings;
use clutch_wallet_lib::utils::wallet_lib::WalletLib;

#[derive(Debug, Clone)]
pub struct AppState {
    pub settings: Settings,
    pub database: DatabaseConnection,
    pub wallet_lib: WalletLib,
}

pub async fn db_connect(connection_url: String) -> DatabaseConnection {
    Database::connect(connection_url)
        .await
        .expect("failed to connect to database")
}
