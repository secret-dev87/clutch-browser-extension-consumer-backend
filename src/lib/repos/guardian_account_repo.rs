use sea_orm::Set;
use sea_orm::{entity::prelude::*, sea_query::Expr};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Serialize, Deserialize)]
#[sea_orm(table_name = "account_guardians")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: String,
    pub guardian_id: String,
    pub account_id: String,
    pub status: String,
}

#[derive(Copy, Clone, Debug, EnumIter)]
pub enum Relation {}

impl RelationTrait for Relation {
    fn def(&self) -> RelationDef {
        panic!("No RelationDef")
    }
}

impl ActiveModelBehavior for ActiveModel {}

pub async fn create(
    db: &DatabaseConnection,
    id: Uuid,
    guardian_id: String,
    account_id: String,
    status: String,
) -> anyhow::Result<()> {
    let model = ActiveModel {
        id: Set(id.to_string()),
        guardian_id: Set(guardian_id.to_owned()),
        account_id: Set(account_id.to_owned()),
        status: Set(status.to_owned()),
    };

    Entity::insert(model)
        .exec_without_returning(db)
        .await
        .map(|_| ())
        .map_err(|e| anyhow::anyhow!(e))
}

pub async fn find_all_accounts_by_guardian_id(
    db: &DatabaseConnection,
    guardian_id: String,
) -> anyhow::Result<Vec<Model>> {
    Entity::find()
        .filter(Column::GuardianId.eq(guardian_id))
        .all(db)
        .await
        .map_err(|e| anyhow::anyhow!(e))
}

pub async fn find_guardian_by_guardian_id_and_account_id(
    db: &DatabaseConnection,
    guardian_id: String,
    account_id: String,
) -> anyhow::Result<Option<Model>> {
    Entity::find()
        .filter(Column::GuardianId.eq(guardian_id))
        .filter(Column::AccountId.eq(account_id))
        .one(db)
        .await
        .map_err(|e| anyhow::anyhow!(e))
}

pub async fn find_all_accounts_by_guardian_id_and_account_id(
    db: &DatabaseConnection,
    guardian_id: String,
    account_id: String,
) -> anyhow::Result<Vec<Model>> {
    Entity::find()
        .filter(Column::GuardianId.eq(guardian_id))
        .filter(Column::AccountId.eq(account_id))
        .all(db)
        .await
        .map_err(|e| anyhow::anyhow!(e))
}

pub async fn find_all_guardians_by_account_id(
    db: &DatabaseConnection,
    account_id: String,
) -> anyhow::Result<Vec<Model>> {
    Entity::find()
        .filter(Column::AccountId.eq(account_id))
        .all(db)
        .await
        .map_err(|e| anyhow::anyhow!(e))
}

pub async fn find_all_guardians_by_account_id_and_status(
    db: &DatabaseConnection,
    account_id: String,
    status: String,
) -> anyhow::Result<Vec<Model>> {
    Entity::find()
        .filter(Column::AccountId.eq(account_id))
        .filter(Column::Status.eq(status))
        .all(db)
        .await
        .map_err(|e| anyhow::anyhow!(e))
}

pub async fn find_all_active_guardians_by_account_id(
    db: &DatabaseConnection,
    account_id: String,
) -> anyhow::Result<Vec<Model>> {
    find_all_guardians_by_account_id_and_status(db, account_id, "ACTIVE".to_owned()).await
}

pub async fn delete_by_id(db: &DatabaseConnection, id: String) -> anyhow::Result<()> {
    Entity::delete_many()
        .filter(Column::Id.eq(id))
        .exec(db)
        .await
        .map_err(|e| anyhow::anyhow!(e))
        .map(|_| ())
}

pub async fn find_all_guardians_for_account_by_ids(
    db: &DatabaseConnection,
    ids: Vec<String>,
    account_id: String,
) -> anyhow::Result<Vec<Model>> {
    Entity::find()
        .filter(Column::AccountId.eq(account_id))
        .filter(Column::Id.is_in(ids))
        .all(db)
        .await
        .map_err(|e| anyhow::anyhow!(e))
}

pub async fn update_all_guardians_for_account_to_status(
    db: &DatabaseConnection,
    account_id: String,
    status: String,
) -> anyhow::Result<()> {
    Entity::update_many()
        .filter(Column::AccountId.eq(account_id))
        .col_expr(Column::Status, Expr::value(status.to_owned()))
        .exec(db)
        .await
        .map_err(|e| anyhow::anyhow!(e))
        .map(|_| ())
}

pub async fn update_guardians_for_account_to_status(
    db: &DatabaseConnection,
    account_id: String,
    ids: Vec<String>,
    status: String,
) -> anyhow::Result<()> {
    Entity::update_many()
        .filter(Column::AccountId.eq(account_id))
        .filter(Column::Id.is_in(ids))
        .col_expr(Column::Status, Expr::value(status.to_owned()))
        .exec(db)
        .await
        .map_err(|e| anyhow::anyhow!(e))
        .map(|_| ())
}
