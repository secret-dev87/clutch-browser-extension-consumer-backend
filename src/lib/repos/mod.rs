pub mod account_repo;
pub mod db;
pub mod guardian_account_repo;
pub mod guardian_repo;
pub mod guardian_settings_repo;
pub mod migration;
pub mod nomination_repo;
pub mod verification_repo;
