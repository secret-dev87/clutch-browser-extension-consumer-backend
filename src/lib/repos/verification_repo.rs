use sea_orm::entity::prelude::*;
use sea_orm::Set;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Serialize, Deserialize)]
#[sea_orm(table_name = "verifications")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: String,
    pub email: String,
    pub code: String,
    pub expires_at: i64,
}

#[derive(Copy, Clone, Debug, EnumIter)]
pub enum Relation {}

impl RelationTrait for Relation {
    fn def(&self) -> RelationDef {
        panic!("No RelationDef")
    }
}

impl ActiveModelBehavior for ActiveModel {}

pub async fn create(
    db: &DatabaseConnection,
    id: Uuid,
    email: &str,
    code: &str,
    expires_at: i64,
) -> anyhow::Result<()> {
    let model = ActiveModel {
        id: Set(id.to_string()),
        email: Set(email.to_owned()),
        code: Set(code.to_owned()),
        expires_at: Set(expires_at.to_owned()),
    };

    Entity::insert(model)
        .exec_without_returning(db)
        .await
        .map(|_| ())
        .map_err(|e| anyhow::anyhow!(e))
}

pub async fn count_by_email(db: &DatabaseConnection, email: &str) -> anyhow::Result<u64> {
    Entity::find()
        .filter(Column::Email.eq(email))
        .count(db)
        .await
        .map_err(|e| anyhow::anyhow!(e))
}

pub async fn find_all_by_email(
    db: &DatabaseConnection,
    email: String,
) -> anyhow::Result<Vec<Model>> {
    Entity::find()
        .filter(Column::Email.eq(email))
        .all(db)
        .await
        .map_err(|e| anyhow::anyhow!(e))
}

pub async fn find_by_email_and_code(
    db: &DatabaseConnection,
    email: &str,
    code: &str,
) -> anyhow::Result<Option<Model>> {
    Entity::find()
        .filter(Column::Email.eq(email))
        .filter(Column::Code.eq(code))
        .one(db)
        .await
        .map_err(|e| anyhow::anyhow!(e))
}
