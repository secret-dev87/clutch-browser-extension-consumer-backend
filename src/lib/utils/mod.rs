use ethers::types::H160;

pub fn convert_to_hex(addr: H160) -> String {
    format!("0x{}", hex::encode(addr))
}
