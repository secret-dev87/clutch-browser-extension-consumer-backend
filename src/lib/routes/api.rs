use crate::repos::db::AppState;
use crate::routes::verification_api;
use axum::{response::Html, routing::get, Router};
use hyper::{StatusCode, Uri};

use super::{account_api, guardian_api};

pub fn router(app_state: AppState) -> Router {
    Router::new()
        .route("/", get(handler))
        .nest("/email", verification_api::routes(&app_state))
        .nest("/accounts", account_api::routes(&app_state))
        .nest("/guardian", guardian_api::routes(&app_state))
        .fallback(fallback)
}

async fn handler() -> Html<&'static str> {
    Html("<h1>Hello, World!</h1>")
}

async fn fallback(uri: Uri) -> (StatusCode, String) {
    (StatusCode::NOT_FOUND, format!("No route for {}", uri))
}
