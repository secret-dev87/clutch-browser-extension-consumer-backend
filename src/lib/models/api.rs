use clutch_wallet_lib::utils::{bundler::UserOperationTransport, wallet_lib::Transaction};
use clutch_wallet_lib::utils::wallet_lib::PreFund;
use ethers::types::{U256, Address};
use sea_orm::entity::prelude::*;
use serde::{de, Deserialize, Deserializer, Serialize};
use std::{fmt, str::FromStr};
use utoipa::ToSchema;

#[derive(Serialize, Deserialize, Debug)]
pub enum ApiPayload<S, E> {
    Success(S),
    Error(E),
}

#[derive(Serialize, Deserialize, Debug)]
pub struct ApiResponse<T, E> {
    pub status: ApiResponseStatus,
    pub payload: ApiPayload<T, E>,
}

pub fn api_error<T>(error_message: String) -> ApiResponse<T, ApiErrorResponse> {
    ApiResponse {
        status: ApiResponseStatus::Error,
        payload: ApiPayload::Error(ApiErrorResponse { error_message }),
    }
}

pub fn api_success<T>(payload: T) -> ApiResponse<T, ApiErrorResponse> {
    ApiResponse {
        status: ApiResponseStatus::Success,
        payload: ApiPayload::Success(payload),
    }
}

#[derive(Serialize, Deserialize, Debug)]
pub enum ApiResponseStatus {
    Success,
    Error,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct ApiErrorResponse {
    pub error_message: String,
}

// Email Verification API
#[derive(Serialize, Deserialize, Debug, Default, ToSchema)]
#[serde(default)]
pub struct VerificationRequest {
    pub email: String,
}

#[derive(Serialize, Deserialize, Debug, Default, ToSchema)]
#[serde(default)]
pub struct VerificationResponse {
    pub success: bool,
}

// Account API
#[derive(Serialize, Deserialize, Debug, Default, ToSchema)]
#[serde(default)]
pub struct AccountCreateRequest {
    pub email: String,
    pub code: String,
    pub paymaster_tokens: Option<Vec<String>>
}

#[derive(Serialize, Deserialize, Debug, Default, ToSchema)]
#[serde(default)]
pub struct AccountCreateResponse {
    pub jwt: String,
    pub contract_wallet_addr: String,
}


#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct FormatUserOpRequest {
    pub max_fee_per_gas: U256,
    pub max_priority_fee_per_gas: U256,
    pub selected_address: Address,
    pub raw_txs: Vec<Transaction>,
    pub pay_token: Address
}

#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(default)]
pub struct FormatUserOpResponse {
    pub user_op: UserOperationTransport,
    pub prefund: PreFund
}


#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(default)]
pub struct AccountUpdateRequest {
    pub email: String,
    pub wallet_address: Option<String>,
    pub eoa_address: Option<String>,
}

#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(default)]
pub struct AccountUpdateResponse {
    pub updated: bool,
}

#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(default)]
#[derive(ToSchema)]
pub struct Account {
    pub id: String,
    pub email: String,
    pub wallet_address: String,
    pub eoa_address: String,
    pub eoa_private_address: String,
    pub updated_at: i64,
}

#[derive(Serialize, Deserialize, Debug, Default, ToSchema)]
#[serde(default)]
pub struct SendTransactionRequest {
    pub from: String,
    pub to: String,
    pub value: String,
}

#[derive(Serialize, Deserialize, Debug, Default, ToSchema)]
#[serde(default)]
pub struct SendTransactionResponse {
    pub status: String,
}

#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(default)]
pub struct PrefundRequest {
    pub send_type: String,
    pub value: Option<String>,
    pub pay_token: Option<String>,
    pub from: String,
    pub to: String //receiver when eth sending, token contract when erc20 sending
}

#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(default)]
pub struct PrefundResponse {
    pub deposit: String,
    pub prefund: String,
    pub missfund: String,
}

#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(default)]
pub struct ListAccountsResponse {
    pub accounts: Vec<Account>,
}

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
pub struct AccountParams {
    #[serde(default, deserialize_with = "empty_string_as_none")]
    pub wallet_address: Option<String>,
    pub eoa_address: Option<String>,
    pub email: Option<String>,
}

fn empty_string_as_none<'de, D, T>(de: D) -> Result<Option<T>, D::Error>
where
    D: Deserializer<'de>,
    T: FromStr,
    T::Err: fmt::Display,
{
    let opt = Option::<String>::deserialize(de)?;
    match opt.as_deref() {
        None | Some("") => Ok(None),
        Some(s) => FromStr::from_str(s).map_err(de::Error::custom).map(Some),
    }
}

// Nominations API
#[derive(Serialize, Deserialize, Debug, Default)]
pub struct NominationCreateRequest {
    pub email: String,
}

#[derive(Serialize, Deserialize, Debug, Default)]
pub struct NominationCreateResponse {
    pub nomination_id: String,
}

#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(default)]
pub struct Nomination {
    pub id: String,
    pub email: String,
    pub guardian_id: String,
    pub account_id: String,
    pub status: String,
}

#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(default)]
pub struct ListNominationsResponse {
    pub nominations: Vec<Nomination>,
}

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
pub struct NominationParams {
    #[serde(default, deserialize_with = "empty_string_as_none")]
    pub status: Option<String>,
    pub nomination_id: Option<String>,
    pub email: Option<String>,
}

#[derive(Serialize, Deserialize, Debug, Default)]
pub struct NominationDeleteResponse {
    pub nomination_id: String,
}

// Guardian API for Guardians

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
pub struct GuardianNominationParams {
    #[serde(default, deserialize_with = "empty_string_as_none")]
    pub status: Option<String>,
    pub nomination_id: Option<String>,
}

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
pub struct GuardianAccountParams {
    #[serde(default, deserialize_with = "empty_string_as_none")]
    pub account_id: Option<String>,
}

#[derive(Serialize, Deserialize, Debug, Default)]
pub struct NominationUpdateRequest {
    pub status: String,
}

#[derive(Serialize, Deserialize, Debug, Default)]
pub struct NominationUpdateResponse {
    pub nomination_id: String,
    pub status: String,
}

#[derive(Serialize, Deserialize, Debug, Default)]
pub struct ListGuardianAccountsResponse {
    pub accounts: Vec<GuardianAccount>,
}

#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(default)]
pub struct GuardianAccount {
    pub id: String,
    pub email: String,
    pub wallet_address: String,
}

// Guardian API for Accounts

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
pub struct AccountGuardianParams {
    #[serde(default, deserialize_with = "empty_string_as_none")]
    pub guardian_id: Option<String>,
    pub status: Option<String>,
}

#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(default)]
pub struct AccountGuardian {
    pub id: String,
    pub email: String,
    pub wallet_address: String,
    pub status: String,
}

#[derive(Serialize, Deserialize, Debug, Default)]
pub struct ListAccountGuardiansResponse {
    pub guardians: Vec<AccountGuardian>,
}

#[derive(Serialize, Deserialize, Debug, Default)]
pub struct AccountGuardianDeleteResponse {
    pub guardian_id: String,
}

// Guardian Settings API

#[derive(Serialize, Deserialize, Debug)]
pub struct AccountGuardianSettingsResponse {
    pub signing_strategies: Vec<SigningStrategy>,
    pub signers: SigningStrategy,
    pub active_guardians: Vec<AccountGuardian>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct AccountGuardianSettingsRequest {
    pub signers: SigningStrategy,
    pub guardians: Vec<String>,
}

#[derive(EnumIter, DeriveActiveEnum, Debug, Clone, Serialize, Deserialize, PartialEq)]
#[sea_orm(rs_type = "String", db_type = "String(Some(1))")]
pub enum SigningStrategy {
    #[sea_orm(string_value = "OneOfOne")]
    OneOfOne,
    #[sea_orm(string_value = "OneOfTwo")]
    OneOfTwo,
    #[sea_orm(string_value = "TwoOfTwo")]
    TwoOfTwo,
    #[sea_orm(string_value = "OneOfThree")]
    OneOfThree,
    #[sea_orm(string_value = "TwoOfThree")]
    TwoOfThree,
    #[sea_orm(string_value = "ThreeOfThree")]
    ThreeOfThree,
}

impl SigningStrategy {
    pub fn get_signers_for(item: SigningStrategy) -> anyhow::Result<i64> {
        match item {
            SigningStrategy::OneOfOne => Ok(1),
            SigningStrategy::OneOfTwo => Ok(2),
            SigningStrategy::TwoOfTwo => Ok(2),
            SigningStrategy::OneOfThree => Ok(3),
            SigningStrategy::TwoOfThree => Ok(3),
            SigningStrategy::ThreeOfThree => Ok(3),
        }
    }

    pub fn all() -> Vec<SigningStrategy> {
        vec![
            SigningStrategy::OneOfOne,
            SigningStrategy::OneOfTwo,
            SigningStrategy::TwoOfTwo,
            SigningStrategy::OneOfThree,
            SigningStrategy::TwoOfThree,
            SigningStrategy::ThreeOfThree,
        ]
    }
}
