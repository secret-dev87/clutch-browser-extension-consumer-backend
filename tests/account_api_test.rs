use axum::http::StatusCode;
use axum_test_helper::TestClient;
use hyper::header::AUTHORIZATION;
use lib::models::api::AccountCreateResponse;
use lib::models::api::ApiErrorResponse;
use lib::models::api::ApiResponse;
use lib::models::api::ListAccountsResponse;
use lib::operations::time::get_unix_timestamp_ms;
use lib::repos::account_repo;
use lib::repos::verification_repo;
use lib::test::utils::create_account;
use lib::test::utils::create_verified_account_jwt;
use lib::test::utils::create_verify;
use lib::test::utils::setup;
use lib::test::utils::tear_down;
use sea_orm::DatabaseConnection;
use uuid::Uuid;

#[tokio::test]
async fn test_successfully_create_account_with_jwt() {
    let (client, app_state, db_url) = setup().await;

    let email = "someone@example.com".to_string();
    let id = Uuid::new_v4();
    let one_minute = 60 * 1000;
    let expires_at = get_unix_timestamp_ms() + one_minute;
    let code = "123456";

    verification_repo::create(&app_state.database, id, &email, code, expires_at)
        .await
        .expect("error creating verification");

    let res = create_account(&client, email.clone(), code.to_string()).await;
    assert_eq!(res.status(), StatusCode::OK);

    let json_response = res
        .json::<ApiResponse<AccountCreateResponse, ApiErrorResponse>>()
        .await;

    insta::assert_yaml_snapshot!(json_response, {
        ".**.jwt" => "[jwt]"
    });

    tear_down(db_url).await;
}

#[tokio::test]
async fn test_invalid_email_format() {
    let (client, _app_state, db_url) = setup().await;

    let res = client
        .post("/accounts")
        .body("{\"email\":\"a\"}")
        .header("Content-Type", "application/json")
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_yaml_snapshot!(res.text().await);

    tear_down(db_url).await;
}

#[tokio::test]
async fn test_error_when_account_already_exists() {
    let (client, app_state, db_url) = setup().await;

    let email = "someone@example.com".to_string();
    create_verify(&client, email.clone()).await;

    let code = verification_repo::find_all_by_email(&app_state.database, email.clone())
        .await
        .expect("error getting code")
        .first()
        .expect("no code found")
        .code
        .clone();

    let created = create_account(&client, email.clone(), code.clone()).await;
    assert_eq!(created.status(), StatusCode::OK);

    let res = create_account(&client, email.clone(), code.clone()).await;
    assert_eq!(res.status(), StatusCode::OK);

    insta::assert_yaml_snapshot!(res.text().await);

    tear_down(db_url).await;
}

#[tokio::test]
async fn test_error_when_verification_not_found_for_account() {
    let (client, _app_state, db_url) = setup().await;

    let email = "someone@example.com".to_string();

    let res = create_account(&client, email.clone(), "123456".to_string()).await;
    assert_eq!(res.status(), StatusCode::OK);

    insta::assert_yaml_snapshot!(res.text().await);

    tear_down(db_url).await;
}

#[tokio::test]
async fn test_error_when_verification_code_expired() {
    let (client, app_state, db_url) = setup().await;

    let email = "someone@example.com".to_string();
    let id = Uuid::new_v4();
    let one_minute = 60 * 1000;
    let expires_at = get_unix_timestamp_ms() - one_minute;
    let code = "123456";

    verification_repo::create(&app_state.database, id, &email, code, expires_at)
        .await
        .expect("error creating verification");

    let res = create_account(&client, email.clone(), code.to_string()).await;
    assert_eq!(res.status(), StatusCode::OK);

    insta::assert_yaml_snapshot!(res.text().await);

    tear_down(db_url).await;
}

#[tokio::test]
async fn test_retrieve_all_accounts() {
    let (client, app_state, db_url) = setup().await;

    let email = "first@example.com".to_string();
    create_verified_account(
        &app_state.database,
        &client,
        email,
        "".to_string(),
        "".to_string(),
    )
    .await;
    let res = client.get("/accounts").send().await;
    assert_eq!(res.status(), StatusCode::OK);

    let json_response = res
        .json::<ApiResponse<ListAccountsResponse, ApiErrorResponse>>()
        .await;

    insta::assert_yaml_snapshot!(json_response, {
        ".**.accounts[0].id" => "[uuid]",
        ".**.accounts[0].updated_at" => "[timestamp]"
    });

    tear_down(db_url).await;
}

#[tokio::test]
async fn test_find_all_by_wallet_address() {
    let (client, app_state, db_url) = setup().await;

    let email = "first@example.com".to_string();
    let wallet_address = "123".to_string();
    create_verified_account(
        &app_state.database,
        &client,
        email,
        wallet_address.clone(),
        "".to_string(),
    )
    .await;
    create_verified_account(
        &app_state.database,
        &client,
        "another@me.com".to_string(),
        "another_wallet".to_string(),
        "".to_string(),
    )
    .await;
    let res = client
        .get(format!("/accounts?wallet_address={}", wallet_address).as_str())
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);

    let json_response = res
        .json::<ApiResponse<ListAccountsResponse, ApiErrorResponse>>()
        .await;

    insta::assert_yaml_snapshot!(json_response, {
        ".**.accounts[0].id" => "[uuid]",
        ".**.accounts[0].updated_at" => "[timestamp]"
    });

    tear_down(db_url).await;
}

#[tokio::test]
async fn test_find_all_by_eoa_address() {
    let (client, app_state, db_url) = setup().await;

    let email = "first@example.com".to_string();
    let eao_address = "123".to_string();
    create_verified_account(
        &app_state.database,
        &client,
        email,
        "".to_string(),
        eao_address.clone(),
    )
    .await;
    create_verified_account(
        &app_state.database,
        &client,
        "another@me.com".to_string(),
        "".to_string(),
        "another_eoa".to_string(),
    )
    .await;
    let res = client
        .get(format!("/accounts?eoa_address={}", eao_address).as_str())
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);

    let json_response = res
        .json::<ApiResponse<ListAccountsResponse, ApiErrorResponse>>()
        .await;

    insta::assert_yaml_snapshot!(json_response, {
        ".**.accounts[0].id" => "[uuid]",
        ".**.accounts[0].updated_at" => "[timestamp]"
    });

    tear_down(db_url).await;
}

#[tokio::test]
async fn test_find_all_by_email_address() {
    let (client, app_state, db_url) = setup().await;

    let email = "first@example.com".to_string();

    create_verified_account(
        &app_state.database,
        &client,
        email.clone(),
        "".to_string(),
        "".to_string(),
    )
    .await;
    create_verified_account(
        &app_state.database,
        &client,
        "another@me.com".to_string(),
        "".to_string(),
        "".to_string(),
    )
    .await;
    let res = client
        .get(format!("/accounts?email={}", email.clone()).as_str())
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);

    let json_response = res
        .json::<ApiResponse<ListAccountsResponse, ApiErrorResponse>>()
        .await;

    insta::assert_yaml_snapshot!(json_response, {
        ".**.accounts[0].id" => "[uuid]",
        ".**.accounts[0].updated_at" => "[timestamp]"
    });

    tear_down(db_url).await;
}

#[tokio::test]
async fn test_successfully_update_account() {
    let (client, app_state, db_url) = setup().await;

    let email = "first@example.com".to_string();
    let wallet_address = "123".to_string();
    let jwt = create_verified_account_jwt(&app_state.database, &client, email.clone()).await;

    let res = client
        .put("/accounts")
        .body(format!(
            "{{\"email\":\"{}\",\"wallet_address\":\"{}\"}}",
            email.clone(),
            wallet_address
        ))
        .header("Content-Type", "application/json")
        .header(AUTHORIZATION, format!("Bearer {}", jwt))
        .send()
        .await;

    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_yaml_snapshot!(res.text().await);

    tear_down(db_url).await;
}

#[tokio::test]
async fn test_error_when_no_bearer_token_on_update_account() {
    let (client, app_state, db_url) = setup().await;

    let email = "first@example.com".to_string();
    let wallet_address = "123".to_string();
    create_verified_account(
        &app_state.database,
        &client,
        email.clone(),
        wallet_address.clone(),
        "".to_string(),
    )
    .await;

    let res = client
        .put("/accounts")
        .body(format!(
            "{{\"email\":\"{}\",\"wallet_address\":\"{}\"}}",
            email.clone(),
            wallet_address
        ))
        .header("Content-Type", "application/json")
        .send()
        .await;

    assert_eq!(res.status(), StatusCode::BAD_REQUEST);
    insta::assert_yaml_snapshot!(res.text().await);

    tear_down(db_url).await;
}

#[tokio::test]
async fn test_error_when_cannot_decode_bearer_token_on_update_account() {
    let (client, app_state, db_url) = setup().await;

    let email = "first@example.com".to_string();
    let wallet_address = "123".to_string();
    create_verified_account(
        &app_state.database,
        &client,
        email.clone(),
        wallet_address.clone(),
        "".to_string(),
    )
    .await;

    let res = client
        .put("/accounts")
        .body(format!(
            "{{\"email\":\"{}\",\"wallet_address\":\"{}\"}}",
            email.clone(),
            wallet_address
        ))
        .header("Content-Type", "application/json")
        .header(AUTHORIZATION, "Bearer 123")
        .send()
        .await;

    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_yaml_snapshot!(res.text().await);

    tear_down(db_url).await;
}

// Helper functions

async fn create_verified_account(
    db: &DatabaseConnection,
    client: &TestClient,
    email: String,
    wallet_address: String,
    eoa_addres: String,
) -> (Uuid, String, i64) {
    let res = create_verify(client, email.clone()).await;
    assert_eq!(res.status(), StatusCode::OK);

    let email = email;
    let verify_id = Uuid::new_v4();
    let one_minute = 60 * 1000;
    let verify_expires_at = get_unix_timestamp_ms() + one_minute;
    let code = "123456";

    let account_id = Uuid::new_v4();
    let account_created_at = get_unix_timestamp_ms();

    verification_repo::create(&db, verify_id, &email, code, verify_expires_at)
        .await
        .expect("error creating verification");

    account_repo::create(
        &db,
        account_id,
        email.clone(),
        wallet_address,
        eoa_addres,
        account_created_at,
    )
    .await
    .expect("error creating account");

    (verify_id, code.to_string(), verify_expires_at)
}
