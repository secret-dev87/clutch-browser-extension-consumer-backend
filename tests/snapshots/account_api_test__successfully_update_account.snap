---
source: tests/account_api_test.rs
expression: res.text().await
---
"{\"status\":\"Success\",\"payload\":{\"Success\":{\"updated\":true}}}"

